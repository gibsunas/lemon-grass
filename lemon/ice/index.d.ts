export * from "./types/crust";
export * from "./types/route";
export * from "./types/serviceLocator";
