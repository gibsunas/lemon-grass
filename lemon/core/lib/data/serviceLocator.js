"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ServiceLocator {
    constructor(options) {
        const serviceList = new Map();
        const verbose = options && options.verbose;
        if (verbose) {
            console.log('Verbose Mode');
        }
        this.byKey = (key) => serviceList.get(key);
        this.install = (input) => {
            if (serviceList.has(input.key)) {
                if (verbose) {
                    console.log(`Cannot Install Duplicate KEY ${input.key}`);
                }
                return false;
            }
            if (verbose) {
                console.log(`Installing ${input.key}...`);
            }
            serviceList.set(input.key, input.method);
            return true;
        };
    }
}
exports.ServiceLocator = ServiceLocator;
;
exports.default = ServiceLocator;
