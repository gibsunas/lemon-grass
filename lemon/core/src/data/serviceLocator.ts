import { 
    HasKey,
    HasMethod,
    Installable,
    Locator,
    LocatorOptions
} from '@lemon/ice';

export class ServiceLocator implements Locator {
    private serviceList:Map<string, any>;
    public byKey: (key:string) => any | undefined;
    public install: (input:Installable) => boolean;
    constructor(options?: LocatorOptions) {
        const serviceList = new Map();

        const verbose = options && options.verbose;
        if(verbose) {console.log('Verbose Mode');}

        this.byKey = (key:string):undefined | any => {
            const exists:boolean = serviceList.has(key);
            if(!exists) { return undefined };

            return serviceList.get(key).method ? serviceList.get(key).method : serviceList.get(key);
        };
        this.install = (input:Installable) => {
            if(serviceList.has(input.key)) {
                if(verbose) { console.log(`Cannot Install Duplicate KEY ${(input as HasKey).key}`);}
                return false;
            }
            if(verbose) { console.log(`Installing ${input.key}...`);}
            serviceList.set(input.key, input.method);

            return true;
        };
    }
};

export default ServiceLocator;