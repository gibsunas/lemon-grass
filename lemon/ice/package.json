{
  "name":"@lemon/ice",
  "module":"./index.ts",
  "description": "A collection of typings",
  "dependencies": {},
  "version":"1.0.0",
  "scripts": {
    "tsc": "tsc --module commonjs"
  }

}
