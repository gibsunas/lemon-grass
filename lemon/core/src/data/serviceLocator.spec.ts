import { LocatorOptions } from "@lemon/ice";

import {ServiceLocator} from "./serviceLocator";

const emptyService = {key:'empty', method:()=>{}};

test("Service Locator Constucts",() => {
    const service = new ServiceLocator();
    expect(service).toBeTruthy();

    // Has expected methods
    expect(service.byKey).toBeTruthy();
    expect(service.install).toBeTruthy();

    expect(service).toMatchSnapshot();
});

test("Service Locator test duplicate service install protection",() => {
    const service = new ServiceLocator();
    expect(service.install(emptyService)).toBeTruthy();
    expect(service.install(emptyService)).toBeFalsy();
});
test("Service Locator install empty service",() => {
    const service = new ServiceLocator();
    service.install(emptyService);
    expect(service.byKey("bad-key")).toBeFalsy();
    expect(service.byKey("empty")).toBeTruthy();
});

test("Service Locator verbose mode",() => {
    const service = new ServiceLocator({verbose:true} as LocatorOptions);
});