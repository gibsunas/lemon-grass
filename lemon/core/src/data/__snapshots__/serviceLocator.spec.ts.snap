// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Locator Constucts 1`] = `
ServiceLocator {
  "byKey": [Function],
  "install": [Function],
}
`;
